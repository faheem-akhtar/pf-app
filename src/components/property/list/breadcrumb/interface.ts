export interface BreadcrumbInterface {
  /**
   * Name
   * @example Apartments
   */
  name: string;

  /**
   * Routing parameters
   * @example /en/search?t=1&c=1
   */
  link: string;

  /**
   * Total count for the link
   * @example 4881
   */
  count: number;
}
