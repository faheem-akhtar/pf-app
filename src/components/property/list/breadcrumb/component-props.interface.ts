import { PropertyListBreadcrumbPropsBaseInterface } from './props-base.interface';

export interface PropertyListBreadcrumbComponentPropsInterface extends PropertyListBreadcrumbPropsBaseInterface {}
