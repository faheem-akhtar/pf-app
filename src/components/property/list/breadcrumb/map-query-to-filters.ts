import { FiltersParametersEnum } from 'enums/filters/parameters.enum';
import { FiltersQueryParametersEnum } from 'enums/filters/query-parameters.enum';
import { LocationService } from 'services/location/service';

const entries: string[] = Object.values(FiltersQueryParametersEnum);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type MapQueryToFiltersResultType = { [key in FiltersParametersEnum]?: any };

/**
 * @param link Clicked breadcrumb link
 */
export const breadcrumbMapQueryToFilters = (link: string): MapQueryToFiltersResultType => {
  const query = link.split('search?')[1];
  const values: MapQueryToFiltersResultType = {};

  query.split('&').forEach((param) => {
    const [queryKey, queryValue] = param.split('=');
    const indexOfParam = entries.indexOf(queryKey);
    const filtersQueryParameterKey = Object.keys(FiltersQueryParametersEnum)[indexOfParam];

    if (queryKey === FiltersQueryParametersEnum.locationsIds) {
      values[FiltersParametersEnum.locationsIds] = queryValue
        .split('-')
        ?.map((id) => LocationService.find((location) => location.id === id));
      return;
    }

    values[FiltersParametersEnum[filtersQueryParameterKey as keyof typeof FiltersParametersEnum]] = queryValue;
  });

  return values;
};
