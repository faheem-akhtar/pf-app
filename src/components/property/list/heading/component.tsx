import { WrapperTemplate } from 'components/wrapper/template';

import styles from '../property-list.module.scss';

export const PropertyListHeadingComponent = ({ pageTitle }: { pageTitle: string }): JSX.Element => {
  const title = pageTitle.split('-')[0];

  return (
    <WrapperTemplate className={styles.heading}>
      <h1 className={styles.title}>{title}</h1>
    </WrapperTemplate>
  );
};
