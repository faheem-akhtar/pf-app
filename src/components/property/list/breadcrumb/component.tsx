import { useRouter } from 'next/router';
import { FunctionComponent, useContext } from 'react';

import { FiltersContext } from 'components/filters/context';
import { FiltersValueInterface } from 'components/filters/value/interface';

import { PropertyListBreadcrumbComponentPropsInterface } from './component-props.interface';
import { breadcrumbMapQueryToFilters } from './map-query-to-filters';
import { PropertyListBreadcrumbTemplate } from './template';

export const PropertyListBreadcrumbComponent: FunctionComponent<PropertyListBreadcrumbComponentPropsInterface> = (
  props
) => {
  const locale = useRouter().locale as string;
  const filtersCtx = useContext(FiltersContext);

  const onClick = (): FiltersValueInterface =>
    filtersCtx.change((filtersValue) => ({
      ...filtersValue,
      ...breadcrumbMapQueryToFilters(locale, props.link),
    }));

  return <PropertyListBreadcrumbTemplate {...props} onClick={onClick} />;
};
