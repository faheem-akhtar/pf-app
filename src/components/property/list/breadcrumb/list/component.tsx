import { FunctionComponent } from 'react';

import { IconThinChevronDownTemplate } from 'components/icon';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';

import { PropertyListBreadcrumbComponent } from '../component';
import styles from '../property-list-breadcrumb.module.scss';
import { PropertyListBreadcrumbListComponentPropsInterface } from './component-props.interface';
import { propertyListBreadcrumbListItemsVisibleCount } from './items-visible-count';

export const PropertyListBreadcrumbListComponent: FunctionComponent<PropertyListBreadcrumbListComponentPropsInterface> =
  ({ breadcrumbs, onClickShowMore, expanded, t }) => (
    <section role='section' className={styles.breadcrumb}>
      <ul role='list' className={styles.breadcrumb__list}>
        {breadcrumbs.slice(0, propertyListBreadcrumbListItemsVisibleCount).map((breadcrumb) => (
          <PropertyListBreadcrumbComponent key={breadcrumb.name} {...breadcrumb} />
        ))}

        {breadcrumbs.length > propertyListBreadcrumbListItemsVisibleCount && (
          <li role='listitem'>
            <ButtonTemplate
              className={styles.breadcrumb__button}
              type='button'
              componentType={ButtonComponentTypeEnum.secondaryBlue}
              size={ButtonSizeEnum.small}
              onClick={onClickShowMore}
              icon={{
                component: IconThinChevronDownTemplate,
                position: ButtonIconPositionEnum.right,
                className: domClassMerge(styles.breadcrumb__icon, {
                  [styles['breadcrumb__icon--expanded']]: expanded,
                }),
              }}
            >
              {t(expanded ? 'show-less' : 'show-more')}
            </ButtonTemplate>
          </li>
        )}
      </ul>

      <ul
        className={domClassMerge(styles.breadcrumb__list, styles['breadcrumb__list--more'], {
          [styles['breadcrumb__list--expanded']]: expanded,
        })}
      >
        {breadcrumbs.slice(propertyListBreadcrumbListItemsVisibleCount, breadcrumbs.length).map((breadcrumb) => (
          <PropertyListBreadcrumbComponent
            key={breadcrumb.name}
            class={styles['breadcrumb__item--more']}
            {...breadcrumb}
          />
        ))}
      </ul>
    </section>
  );
