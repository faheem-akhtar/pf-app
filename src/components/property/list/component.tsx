import Head from 'next/head';
import React, { useContext, useRef } from 'react';

import { AuthModalComponent } from 'components/auth/modal/component';
import { AuthScreenEnum } from 'components/auth/screen.enum';
import { ModalComponent } from 'components/modal/component';
import { PropertyCardComponent } from 'components/property-card/component';
import { SavedPropertyAuthLoginTemplate } from 'components/saved-property/auth/login/template';
import { SAVED_PROPERTY_AUTH_MODAL_STORAGE_KEY } from 'components/saved-property/auth/modal/storage-key.constant';
import { savedPropertyTracker } from 'components/saved-property/tracker';
import { configAdsGptUnits } from 'config/ads/gpt/units';
import { UserContext } from 'context/user/context';
import { functionNoop } from 'helpers/function/noop';
import { useServicesDfpAds } from 'services/dfp/ads.hook';
import { WindowService } from 'services/window/service';

import { PROPERTY_SERP_NO_OF_PRELOAD_IMAGES } from '../serp/no-of-preload-images.constant';
import { propertySerpObfuscatedGetImgUrl } from '../serp/obfuscated/get/img-url';
import { propertySerpObfuscatedGetUrl } from '../serp/obfuscated/get/url';
import { PropertySerpObfuscatedType } from '../serp/obfuscated/type';
import { PropertyTrackerFactory } from '../tracker.factory';
import { PropertyListAdPlaceholderTemplate } from './ad-placeholder-template';
import { PropertyListComponentPropsInterface } from './component-props.interface';
import { PropertyListItemType } from './item/type';
import { PropertyListItemTypeEnum } from './item/type.enum';
import styles from './property-list.module.scss';

export const PropertyListComponent: React.FunctionComponent<PropertyListComponentPropsInterface> = (props) => {
  const { properties, adConfig, pageIsLoading } = props;
  const { sessionStorage } = WindowService;
  const propertyTracker = PropertyTrackerFactory();
  const user = useContext(UserContext);

  const authModalOpenRef = useRef<() => void>(functionNoop);
  const authModalCloseRef = useRef<() => void>(functionNoop);

  const items: PropertyListItemType[] = properties.map((property: PropertySerpObfuscatedType) => ({
    type: PropertyListItemTypeEnum.property,
    property,
  }));

  configAdsGptUnits.forEach((unit) => {
    if (unit.position !== undefined) {
      items.splice(unit.position, 0, { type: PropertyListItemTypeEnum.ad, id: unit.id });
    }
  });

  useServicesDfpAds(adConfig);

  const PreloadImages = (): JSX.Element => (
    <Head>
      {properties.slice(0, PROPERTY_SERP_NO_OF_PRELOAD_IMAGES).map((p, index) => (
        <link
          key={index}
          rel='preload'
          href={propertySerpObfuscatedGetImgUrl(p).replace(/\.jpg\b/, '.webp')}
          as='image'
        />
      ))}
    </Head>
  );

  return (
    <>
      <PreloadImages />
      <div className={styles.container}>
        {items.map((item, index) => {
          if (item.type === 'ad') {
            return <PropertyListAdPlaceholderTemplate id={item.id} key={index} />;
          }

          return (
            <PropertyCardComponent
              key={`${propertySerpObfuscatedGetUrl(item.property)}-${index}`}
              property={item.property}
              loading={pageIsLoading}
              onSaveButtonClick={(propertyId, isSaved): void => {
                if (isSaved && !sessionStorage.getItem(SAVED_PROPERTY_AUTH_MODAL_STORAGE_KEY) && !user) {
                  authModalOpenRef.current();
                  sessionStorage.setItem(SAVED_PROPERTY_AUTH_MODAL_STORAGE_KEY, { propertyId });
                  savedPropertyTracker.onOpenUserAuthModal();
                }
              }}
              onPropertyClick={(): void => propertyTracker.click(item.property)}
            />
          );
        })}
        <ModalComponent openRef={authModalOpenRef} closeRef={authModalCloseRef} overlay>
          <AuthModalComponent
            eventLabel='Property Serp - Post Call Lead'
            initialScreen={AuthScreenEnum.shortLogin}
            loginTemplate={SavedPropertyAuthLoginTemplate}
            close={(): void => authModalCloseRef.current()}
          />
        </ModalComponent>
      </div>
    </>
  );
};
