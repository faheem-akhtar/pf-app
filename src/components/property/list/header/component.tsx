import { FunctionComponent, useState } from 'react';

import { WrapperTemplate } from 'components/wrapper/template';
import { useTranslation } from 'helpers/translation/hook';

import { PropertyListBreadcrumbListComponent } from '../breadcrumb/list/component';
import { PropertyListHeaderComponentPropsInterface } from './component-props.interface';
import styles from './property-list-header.module.scss';

export const PropertyListHeaderComponent: FunctionComponent<PropertyListHeaderComponentPropsInterface> = (props) => {
  const { t } = useTranslation();
  const [isExpanded, setIsExpanded] = useState<boolean>(false);

  const pageTitle = props.pageTitle.split('-')[0];

  return (
    <WrapperTemplate className={styles.heading}>
      <>
        <h1 className={styles.title}>{pageTitle}</h1>

        {props.breadcrumbs.length ? (
          <PropertyListBreadcrumbListComponent
            breadcrumbs={props.breadcrumbs}
            onClickShowMore={(): void => setIsExpanded((isExpanded) => !isExpanded)}
            expanded={isExpanded}
            t={t}
          />
        ) : null}
      </>
    </WrapperTemplate>
  );
};
