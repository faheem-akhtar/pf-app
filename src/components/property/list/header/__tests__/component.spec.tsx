import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { filtersContextPropsStub } from 'stubs/filters/context-props.stub';
import { propertyListBreadcrumbStub } from 'stubs/property/list/breadcrumb/stub';

import { FiltersContextProvider } from 'components/filters/context-provider';

import { PropertyListHeaderComponent } from '../component';

const breadcrumbStub = [
  {
    name: 'Hotel Apartments',
    link: '/en/search?t=1111&c=1',
    count: 1,
  },
];

describe('PropertyListHeaderComponent', () => {
  const pageTitle = 'Apartments for sale in UAE - 41886 Flats for sale | Property Finder UAE';

  it('renders without throwing any errors', () => {
    const { container } = render(
      <FiltersContextProvider {...filtersContextPropsStub()}>
        <PropertyListHeaderComponent pageTitle={pageTitle} breadcrumbs={propertyListBreadcrumbStub()} />
      </FiltersContextProvider>
    );

    expect(container).toMatchSnapshot();
  });

  test('if there are breadcrumbs data', () => {
    render(
      <FiltersContextProvider {...filtersContextPropsStub()}>
        <PropertyListHeaderComponent pageTitle={pageTitle} breadcrumbs={propertyListBreadcrumbStub(breadcrumbStub)} />
      </FiltersContextProvider>
    );

    userEvent.click(screen.getByText('show-more'));

    expect(screen.getByText('Apartments')).toBeInTheDocument();
  });

  it('should not visible breadcrumbs when no data', () => {
    render(
      <FiltersContextProvider {...filtersContextPropsStub()}>
        <PropertyListHeaderComponent pageTitle={pageTitle} breadcrumbs={[]} />
      </FiltersContextProvider>
    );

    expect(screen.queryByRole('section')).not.toBeInTheDocument();
  });

  it('should expanded state work properly when breadcrumb more or less are clicked', () => {
    render(
      <FiltersContextProvider {...filtersContextPropsStub()}>
        <PropertyListHeaderComponent pageTitle={pageTitle} breadcrumbs={propertyListBreadcrumbStub(breadcrumbStub)} />
      </FiltersContextProvider>
    );

    expect(screen.getByText('show-more')).toBeInTheDocument();

    userEvent.click(screen.getByText('show-more'));

    expect(screen.getByText('show-less')).toBeInTheDocument();

    userEvent.click(screen.getByText('show-less'));

    expect(screen.getByText('show-more')).toBeInTheDocument();
  });
});
