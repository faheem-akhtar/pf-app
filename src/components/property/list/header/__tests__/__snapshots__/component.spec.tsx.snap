// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyListHeaderComponent renders without throwing any errors 1`] = `
<div>
  <div
    class="container heading"
  >
    <h1
      class="title"
    >
      Apartments for sale in UAE 
    </h1>
    <section
      class="breadcrumb"
      role="section"
    >
      <ul
        class="breadcrumb__list"
        role="list"
      >
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Apartments
            <span
              class="breadcrumb__count"
            >
              (
              4,879
              )
            </span>
          </a>
        </li>
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Land
            <span
              class="breadcrumb__count"
            >
              (
              894
              )
            </span>
          </a>
        </li>
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Duplexes
            <span
              class="breadcrumb__count"
            >
              (
              445
              )
            </span>
          </a>
        </li>
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Chalets
            <span
              class="breadcrumb__count"
            >
              (
              170
              )
            </span>
          </a>
        </li>
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Villas
            <span
              class="breadcrumb__count"
            >
              (
              160
              )
            </span>
          </a>
        </li>
        <li
          class="breadcrumb__item"
          role="listitem"
        >
          <a
            class="breadcrumb__link"
          >
            Whole Buildings
            <span
              class="breadcrumb__count"
            >
              (
              38
              )
            </span>
          </a>
        </li>
      </ul>
      <ul
        class="breadcrumb__list breadcrumb__list--more"
      />
    </section>
  </div>
</div>
`;
